### PR TITLE
[forge] Add Consensus reliability tests

### PR DESCRIPTION
**Pull request by igor-aptos**
### Test Plan
added 3 continuous tests for consensus reliability

tuning params in stacked PRs

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3958)

### CODE REVIEW

**Review by sitalkedia — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3958)* status: 0 of 16 files reviewed, 10 unresolved discussions (waiting on @bchocho, @davidiw, @igor-aptos, @perryjrandall, @rustielin, @wrwg, and @zekun000)

___
*[`testsuite/forge/src/test_utils/consensus_utils.rs` line 200 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBQaeX5-qp_4xo1PCQ1:-NBQaeX5-qp_4xo1PCQ2:b4p29yg) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/forge/src/test_utils/consensus_utils.rs#L200)):*
> ```XML
>     }
> 
>     if errors.is_empty() {
> ```

if !errors.is_empty() ? 
___
*[`testsuite/forge-cli/src/main.rs` line 539 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBQbUT90LZYCsCuyKL_:-NBQbUT90LZYCsCuyKLa:b-6xfqda) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/forge-cli/src/main.rs#L539)):*
> ```XML
>         ),
>         "large_test_only_few_nodes_down" => changing_working_quorum_test(
>             60,
> ```

Do we need such high number of validators for this test?
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 65 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBQcsj66iKs-kV-3UZw:-NBQcsj66iKs-kV-3UZx:bv27xbe) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L65)):*
> ```XML
>             )?
>             .into_inner();
>         println!("ValidatorSet : {:?}", validator_set);
> ```

let's use the logger instead. 
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 68 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBSY5Ma-JFZg93NgutP:-NBSY5Ma-JFZg93NgutQ:b-mmjwl0) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L68)):*
> ```XML
> 
>         let num_always_healthy = self.always_healthy_nodes;
>         let can_fail_for_quorum = std::cmp::min(
> ```

nit: a comment here explaining the logic would be helpful. 
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 73 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBQcvxLEeJ_PWWsWIbX:-NBQcvxLEeJ_PWWsWIbY:b-7ropln) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L73)):*
> ```XML
>         );
>         let cycle_offset = can_fail_for_quorum / 4 + 1;
>         println!(
> ```

ditto
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 114 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBSYeyH4PSGljSWj5Zi:-NBSYeyH4PSGljSWj5Zj:b-ivihk3) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L114)):*
> ```XML
>                                 (
>                                     *i,
>                                     "consensus::send::any".to_string(),
> ```

Let's add a comment here on explaining the intentions of these fail point.
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 144 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBTUZBPBO9KZ_BY49Ip:-NBTUZBPBO9KZ_BY49Iq:b-s5g3in) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L144)):*
> ```XML
> 
>                 // Make sure that every active node is making progress, so we compare min(cur) vs max(previous)
>                 let epochs = cur_active.iter().map(|s| s.epoch).min().unwrap()
> ```

nit: let's rename this to `epoch_progress` to be clearer



_Code quote:_
```XML
test_consensus_fault_tolerance
```
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 146 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBTUdos86DMI2Ddbbk7:-NBTUdos86DMI2Ddbbk8:b-q53cxw) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L146)):*
> ```XML
>                 let epochs = cur_active.iter().map(|s| s.epoch).min().unwrap()
>                     - previous_active.iter().map(|s| s.epoch).max().unwrap();
>                 let rounds = cur_active
> ```

rename this to `round_prorgress`
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 152 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBTUgyeE4dgeXIZhI7R:-NBTUgyeE4dgeXIZhI7S:bub6tl5) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L152)):*
> ```XML
>                     .unwrap() as i64 -
>                     previous_active.iter().map(|s| s.round).max().unwrap() as i64;
>                 let transactions = cur_active.iter().map(|s| s.version).min().unwrap() as i64
> ```

rename this to `transactions_progress`
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 163 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBTVCNCAh06xdwjIOCB:-NBTVCNCAh06xdwjIOCC:b8kh20o) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L163)):*
> ```XML
>                 }
>                 if epochs == 0 && rounds < (check_period_s / 2) as i64 {
>                     bail!("no progress with active consensus, only {} epochs and {} rounds, expectd >= {}",
> ```

Hmm do we need a check in case of epoch tansition as well? Is it easy to add?

**Review by ikabiljo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3958)* status: 0 of 16 files reviewed, 10 unresolved discussions (waiting on @bchocho, @davidiw, @igor-aptos, @perryjrandall, @rustielin, @sitalkedia, @wrwg, and @zekun000)

___
*[`testsuite/forge/src/test_utils/consensus_utils.rs` line 200 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBQaeX5-qp_4xo1PCQ1:-NBU5w-j9wQPf1YGtgW8:bb7s9zz) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/forge/src/test_utils/consensus_utils.rs#L200)):*
<details><summary><i>Previously, sitalkedia (Sital Kedia) wrote…</i></summary><blockquote>

if !errors.is_empty() ? 
</blockquote></details>

yes!
___
*[`testsuite/forge-cli/src/main.rs` line 539 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBQbUT90LZYCsCuyKL_:-NBU6-QV7kVXz3o7aka7:b-jekez9) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/forge-cli/src/main.rs#L539)):*
<details><summary><i>Previously, sitalkedia (Sital Kedia) wrote…</i></summary><blockquote>

Do we need such high number of validators for this test?
</blockquote></details>

I am not adding this one to the continuous. this is AIT3 like test. We might want to have some larger ones weekly?
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 65 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBQcsj66iKs-kV-3UZw:-NBU6R0uB7MEjoorMIpl:b-qe4n52) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L65)):*
<details><summary><i>Previously, sitalkedia (Sital Kedia) wrote…</i></summary><blockquote>

let's use the logger instead. 
</blockquote></details>

can we make logger print nicer in forge?
___
*[`testsuite/testcases/src/consensus_reliability_tests.rs` line 163 at r1](https://reviewable.io/reviews/aptos-labs/aptos-core/3958#-NBTVCNCAh06xdwjIOCB:-NBU9yKZ8BAbXDQmOTCJ:b-12mm4m) ([raw file](https://github.com/aptos-labs/aptos-core/blob/9f1bb54e81e36f582c99bc0e85d9c6bedf7a232c/testsuite/testcases/src/consensus_reliability_tests.rs#L163)):*
> Hmm do we need a check in case of epoch tansition as well? Is it easy to add?
if there is an epoch change, I assume it is enough (because I don't know how many rounds were there in the previous epoch, before it completed)

File: testsuite/forge-cli/src/main.rs [hunk 1]
```diff
@@ -480,16 +480,19 @@ fn single_test_suite(test_name: &str) -> Result<ForgeConfig<'static>> {
...
             .with_initial_validator_count(NonZeroUsize::new(10).unwrap())
             .with_genesis_helm_config_fn(Arc::new(|helm_values| {
                 helm_values["chain"]["epoch_duration_secs"] = 60.into();
             }))
             .with_node_helm_config_fn(Arc::new(|helm_values| {
-                helm_values["consensus"]["max_block_txns"] = 50.into();
-                helm_values["consensus"]["round_initial_timeout_ms"] = 500.into();
-                helm_values["consensus"]["round_timeout_backoff_exponent_base"] = 1.0.into();
-                helm_values["consensus"]["quorum_store_poll_count"] = 1.into();
+                helm_values["validator"]["config"]["consensus"]["max_block_txns"] = 50.into();
```
Comment (zekun000) [inline testsuite/forge-cli/src/main.rs:489]: so it doesn't work before?

File: testsuite/testcases/src/consensus_reliability_tests.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+        if self.add_execution_delay {
+            runtime.block_on(async {
+                let mut rng = rand::thread_rng();
+                for (name, validator) in &validators[num_always_healthy..num_validators] {
+                    let sleep_time = rng.gen_range(20, 500);
+                    let name = name.clone();
+                    validator
+                        .set_failpoint(
+                            "aptos_vm::execution::block_metadata".to_string(),
```
Comment (zekun000) [inline testsuite/testcases/src/consensus_reliability_tests.rs:100]: is it a stacked diff? I don't see this on main

**Review by zekun000 — APPROVED** [review]

**Review by sitalkedia — APPROVED** [review]
LGTM. As discussed offline, it would be great to follow up with a test for consensus falling back to state sync and ensure the node can catch up and participate in consensus.

File: testsuite/forge-cli/src/main.rs (hunk 1 above)
Comment (igor-aptos) [reply testsuite/forge-cli/src/main.rs:489]: It wasn't applying these (i.e. running round fast). But was still passing as a test

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 3]
```diff
@@ -972,6 +972,7 @@ impl VMAdapter for AptosVM {
     ) -> Result<(VMStatus, TransactionOutputExt, Option<String>), VMStatus> {
         Ok(match txn {
             PreprocessedTransaction::BlockMetadata(block_metadata) => {
+                fail_point!("aptos_vm::execution::block_metadata");
```
Comment (igor-aptos) [inline aptos-move/aptos-vm/src/aptos_vm.rs:975]: @zekun000 - I add it here

**Pull request merged**